dart/common/Addon.cpp

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright (c) 2015, Georgia Tech Research Corporation
+ * All rights reserved.
+ *
+ * Author(s): Michael X. Grey <mxgrey@gatech.edu>
+ *
+ * Georgia Tech Graphics Lab and Humanoid Robotics Lab
+ *
+ * Directed by Prof. C. Karen Liu and Prof. Mike Stilman
+ * <karenliu@cc.gatech.edu> <mstilman@cc.gatech.edu>
+ *
+ * This file is provided under the following "BSD-style" License:
+ *   Redistribution and use in source and binary forms, with or
+ *   without modification, are permitted provided that the following
+ *   conditions are met:
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND
+ *   CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ *   INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ *   MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ *   DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *   CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ *   SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ *   LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
+ *   USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ *   AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *   ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *   POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <cassert>
+#include <string>
+#include <iostream>
+
+#include "dart/common/Addon.h"
+#include "dart/common/Console.h"
+
+namespace dart {
+namespace common {
+
+//==============================================================================
+void Addon::setAddonState(const State& /*otherState*/)
+{
+  // Do nothing
+}
+
+//==============================================================================
+const Addon::State* Addon::getAddonState() const
+{
+  return nullptr;
+}
+
+//==============================================================================
+void Addon::setAddonProperties(const Properties& /*someProperties*/)
+{
+  // Do nothing
+}
+
+//==============================================================================
+const Addon::Properties* Addon::getAddonProperties() const
+{
+  return nullptr;
+}
+
+//==============================================================================
+bool Addon::isOptional(AddonManager* /*oldManager*/)
+{
+  return true;
```

Comment:
Regarding the first point: I couldn't figure out a way of embedding the "optional" information into ``FooManager`` or ``GeneralManager`` such that it could be accessed in constant-time.

Regarding the second point, I think the existing behavior makes sense (although if you disagree, I'm open to discussing it). The ``OptionalT`` template parameter defaults to ``true``, meaning we do not generally expect an Addon to be required. However, if the Addon **is** meant to be required for its particular type of manager (given as the ``ManagerT`` template parameter), then ``AddonWithProtectedPropertiesInSkeleton::isOptional`` will check the manager type to see if it matches ``ManagerT``. If the manager type is some other manager type, then it will return ``true`` for being optional, otherwise it returns ``false`` to indicate that it's a required Addon for that type of manager.

So for example, a ``VisualData`` or ``CollisionData`` addon would have ``OptionalT = true``, indicating that they are optional for any type of manager that they get attached to. Conversely, ``RevoluteJointAddon`` would have ``OptionalT = false`` and ``ManagerT = RevoluteJoint`` indicating that it is required for any manager type that has ``RevoluteJoint`` in its inheritance hierarchy, but it is optional for any other manager type.